### PR TITLE
Migrate makefiles and CI/CD

### PR DESCRIPTION
**Pull request by inteon**
ref: https://github.com/cert-manager/makefile-modules/issues/3

Works as-is, but will be better linked with https://github.com/cert-manager/testing/pull/934/files (adds caching, go vendoring and new target names)

### CODE REVIEW

File: deploy/charts/approver-policy/values.yaml
Comment (wallrj) [inline deploy/charts/approver-policy/values.yaml:114]: This and the change to the CRD template file seems unrelated. Please revert and add it in a separate PR.

File: docs/api/api.md
Comment (wallrj) [inline docs/api/api.md:1]: Why is this so different? Is there a different tool generating this file?

File: .github/workflows/release.yaml
Comment (wallrj) [inline .github/workflows/release.yaml:1]: Do a test release by pushing an alpha tag linked to a commit in this branch to demonstrate that this still works.

Update the README.md file with updated release process documentation, if necessary.

File: .github/workflows/release.yaml
Comment (wallrj) [inline .github/workflows/release.yaml:66]: approver-policy ?

File: .github/workflows/release.yaml
Comment (wallrj) [inline .github/workflows/release.yaml:80]: approver-policy.

File: Makefile
Comment (wallrj) [inline Makefile:16]: How will contributors and maintainers know how to add project specific make targets?
Please add documentation to this file explaining (or linking to an explanation) where to put project-specific make files and how they get included into this file.

File: Makefile
Comment (wallrj) [inline Makefile:77]: `make clean` doesn't seem to be available any more.
How shall I delete all the generated files and start again if I need to?

File: deploy/charts/approver-policy/README.md
Comment (wallrj) [inline deploy/charts/approver-policy/README.md:42]: Revert.

File: deploy/charts/approver-policy/templates/crd-policy.cert-manager.io_certificaterequestpolicies.yaml
Comment (wallrj) [inline deploy/charts/approver-policy/templates/crd-policy.cert-manager.io_certificaterequestpolicies.yaml:1045]: Revert.

File: go.mod
Comment (wallrj) [inline go.mod:3]: Why change this? And why not the latest patch version of Go?

File: go.sum
Comment (wallrj) [inline go.sum:1]: Why did this change?

File: klone.json
Comment (wallrj) [inline klone.json:1]: It's a shame you used `.json` for this because it prevents us adding a comment at the top explaining what this file is for,
how it gets updated and a comment with a link to the `klone` project for further documentation.
Can you change it to yaml?

File: pkg/internal/approver/validation/certificaterequest.pb.go
Comment (wallrj) [inline pkg/internal/approver/validation/certificaterequest.pb.go:4]: Why did this get downgraded?

File: pkg/apis/policy/v1alpha1/zz_generated.deepcopy.go
Comment (wallrj) [inline pkg/apis/policy/v1alpha1/zz_generated.deepcopy.go:4]: Why did this change?

**Review by wallrj — CHANGES_REQUESTED** [review]
Thanks Tim,

This is very cool, but I want to see more documentation in the README of this project and in the README of the upstream project explaining how it works and how to contribute changes.

I feel unsure about checking all this cloned library code in `make/_shared/` into the repo.
It makes it easy to `git grep` to see how everything works.

I checked out the branch and tried some of the new make targets.
Please document the basic `make targets` for contributors in the README file.

I found the `make help` output difficult to read.
This is how `make help` looks for me:

```console
$ make
Usage: make [TARGET [TARGET ...]] [ARGUMENT=VALUE [ARGUMENT=VALUE ...]]
Generate/ Verify:
    generate-protos $(NEEDS_PROTOC) $(NEEDS_PROTOC-GEN-GO)
        Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
Testing:
    test-smoke  $(NEEDS_GINKGO)
        Smoke end-to-end tests
    test-unit $(cert_manager_crds) $(approver_policy_crds) $(NEEDS_GINKGO) $(NEEDS_ETCD) $(NEEDS_KUBE-APISERVER) $(NEEDS_KUBECTL)
        Unit tests
[shared] Build:
$(build_targets): $(bin_dir)/bin/%: FORCE | $(NEEDS_GO) $(bin_dir)/bin
        Build manager binary.
$(oci_build_targets): oci-build-%: | $(NEEDS_KO) $(NEEDS_GO) $(NEEDS_YQ) $(bin_dir)/scratch/image
        Build the oci image.
$(oci_load_targets): oci_platforms := ""
        Load docker image.
$(oci_maybe_push_targets): oci-maybe-push-%: | $(NEEDS_CRANE)
        Push docker image if tag does not already exist.
        Expected pushed images:
        - :v1.2.3, @sha256:0000001
        - :v1.2.3.sig, :sha256-0000001.sig
$(oci_push_targets): oci-push-%: oci-build-% | $(NEEDS_CRANE) $(NEEDS_COSIGN) $(NEEDS_YQ) $(bin_dir)/scratch/image
        Push docker image.
        If the tag already exists, this target will overwrite it.
        If an identical image was already built before, we will add a new tag to it, but we will not sign it again.
        Expected pushed images:
        - :v1.2.3, @sha256:0000001
        - :v1.2.3.sig, :sha256-0000001.sig
$(run_targets): run-%: | $(NEEDS_GO)
        Run a controller from your host.
[shared] Deployment:
    install $(helm_chart_archive) | $(NEEDS_HELM)
        Install controller helm chart on the current active K8S cluster.
    template $(helm_chart_archive) | $(NEEDS_HELM)
        Template the helm chart.
    uninstall $(NEEDS_HELM)
        Uninstall controller helm chart from the current active K8S cluster.
[shared] Generate/ Verify:
    generate-api-docs $(NEEDS_GOMARKDOC)
        Generate API docs for the API types.
    generate-base
        Generate base files in the repository
    generate-crds $(NEEDS_CONTROLLER-GEN)
        Generate CRD manifests.
    generate-deepcopy $(NEEDS_CONTROLLER-GEN)
        Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
    generate-helm-docs $(NEEDS_HELM-DOCS)
        Generate Helm chart documentation.
    generate-klone $(NEEDS_KLONE)
        Generate klone shared Makefiles
    generate $(shared_generate_targets) $(extra_generate_targets)
        Generate all generate targets.
    verify-boilerplate $(NEEDS_BOILERSUITE)
        Verify that all files have the correct boilerplate.
    verify-generate-api-docs $(NEEDS_GOMARKDOC)
        Verify that the API docs are up to date.
    verify-helm-lint $(helm_chart_archive) | $(NEEDS_HELM)
        Verify that the Helm chart is linted.
    verify-pod-security-standards $(helm_chart_archive) | $(NEEDS_KYVERNO) $(NEEDS_KUSTOMIZE) $(NEEDS_HELM)
        Verify that the Helm chart complies with the pod security standards.
    verify $(shared_generate_targets:%=verify-%) $(extra_generate_targets:%=verify-%) $(shared_verify_targets) $(extra_verify_targets)
        Verify code and generate targets.
[shared] Helm Chart:
    helm-chart $(helm_chart_archive)
        Create a helm chart
[shared] Kind cluster:
    images-preload $(images_tar_envs)
        Preload images.
    kind-cluster-clean $(NEEDS_KIND)
        Delete the Kind cluster
    kind-cluster $(kind_kubeconfig) | $(NEEDS_KUBECTL)
        Create Kind cluster and wait for nodes to be ready
    kind-logs  $(NEEDS_KIND) $(bin_dir)/artifacts
        Get the Kind cluster
[shared] Release:
    release $(NEEDS_CRANE)
        Publish all release artifacts (image + helm chart)
[shared] Self-upgrade:
    upgrade-klone $(NEEDS_KLONE)
        Upgrade klone Makefile modules to latest version
[shared] Tools:
    tools-learn-sha
        re-download all tools and learn the sha values, useful after upgrading
    tools $(TOOLS_PATHS) $(K8S_CODEGEN_TOOLS_PATHS)
        Download and setup all tools
    vendor-go $(bin_dir)/tools/go
        By default, this Makefile uses the system's Go. You can use a "vendored"
        version of Go that will get downloaded by running this command once. To
        disable vendoring, run "make unvendor-go". When vendoring is enabled,
        you will want to set the following:

        export PATH="$PWD/$(bin_dir)/tools:$PATH"
        export GOROOT="$PWD/$(bin_dir)/tools/goroot"
    which-go $(NEEDS_GO)
        Print the version and path of go which will be used for building and
        testing in Makefile commands. Vendored go will have a path in ./bin
```

File: deploy/charts/approver-policy/values.yaml
Comment (inteon) [reply deploy/charts/approver-policy/values.yaml:114]: see https://github.com/cert-manager/approver-policy/pull/316/files

File: go.mod
Comment (inteon) [reply go.mod]: I reverted this. I made it so the go version is no longer coupled to the go.mod file.

File: pkg/internal/approver/validation/certificaterequest.pb.go
Comment (inteon) [reply pkg/internal/approver/validation/certificaterequest.pb.go]: Fixed.

File: pkg/apis/policy/v1alpha1/zz_generated.deepcopy.go
Comment (inteon) [reply pkg/apis/policy/v1alpha1/zz_generated.deepcopy.go:4]: I changed the template so it does not include YEAR, and that caused the extra space to be removed.

File: Makefile
Comment (inteon) [reply Makefile:16]: I added some extra comments.

File: klone.json
Comment (inteon) [reply klone.json:1]: Looking into this...

File: .github/workflows/release.yaml
Comment (inteon) [reply .github/workflows/release.yaml:1]: https://github.com/cert-manager/approver-policy/actions/runs/7220836055
https://github.com/cert-manager/approver-policy/releases/tag/untagged-442b7c6d0c2988dd8ca6

File: .github/workflows/release.yaml
Comment (wallrj) [inline .github/workflows/release.yaml:34]: Don't you prefer to point to the `go.mod` file here to automatically build using the latest patch version?
I do.

**Review by wallrj — COMMENTED** [review]
partial review...

File: .github/workflows/release.yaml
Comment (inteon) [reply .github/workflows/release.yaml:34]: The problem is that the go version is not linked to the source code. Instead, it is linked to the latest version of go at that time.
I think for reproducibility and as a way to properly keep track of versions, we should explicitly version go instead.

File: .gitattributes
Comment (inteon) [reply .gitattributes:1]: These changes are part of https://github.com/cert-manager/approver-policy/pull/316.
This PR depends on that PR, because the generation logic for the CRDs also generates the conditional inclusion of the CRDs.

File: klone.json
Comment (inteon) [reply klone.json:1]: Fixed this in the latest edit.

File: .github/workflows/release.yaml
Comment (wallrj) [inline .github/workflows/release.yaml:78]: Great.

File: Makefile
Comment (wallrj) [inline Makefile:29]: Thanks.

File: Makefile
Comment (wallrj) [inline Makefile:77]: Thanks.

File: README.md
Comment (wallrj) [inline README.md:29]: Great.

File: klone.yaml
Comment (wallrj) [reply klone.yaml:1]: Maybe it'd be event nicer to call it `.klone.yaml`, as is customary with other tool config files....but that can be future improvement if you like.

File: klone.yaml
Comment (wallrj) [inline klone.yaml:6]: Perfect!

File: make/02_mod.mk
Comment (wallrj) [inline make/02_mod.mk:46]: TODO now or later?

**Review by wallrj — APPROVED** [review]
This is looking really nice @inteon!

I hadn't appreciated how much work had gone into it.

As far as I can see, this retains all the same tests and greatly improves the release process. Happy for you to merge this now if you prefer to make further improvements in followup PRs.

/lgtm
/hold

File: make/02_mod.mk
Comment (inteon) [reply make/02_mod.mk:46]: This is for later after we change the testing repo.

**Comment by inteon** [comment]
@wallrj I did remove the conditional CRDs from this PR, we will have to fix that later (when we want to use this library with trust-manager).
PTAL, all issues should be resolved now.

**Review by wallrj — APPROVED** [review]
Thanks!

/lgtm
/approve

File: .github/workflows/release.yaml
Comment (wallrj) [reply .github/workflows/release.yaml:34]: Turns out that setup-go doesn't necessarily pick up latest version of Go:

* https://github.com/actions/setup-go/issues/434

**Pull request merged**
